book/src/transaction-fees.md

```diff
@@ -0,0 +1,52 @@
+# Deterministic Transaction Fees
+
+Transactions currently include a fee field that indicates the maximum fee field
+a slot leader is permitted to charge to process a transaction. The cluster, on
+the other hand, agrees on a minimum fee. If the network is congested, the slot
+leader may prioritize the transactions offering higher fees. That means the
+client won't know how much was collected until the transaction is confirmed by
+the cluster and the remaining balance is checked. It smells of exactly we
+dislike about Ethereum's "gas", non-determinism.
+
+## Implementation Status
+
+This design is not yet implemented, but is written as though it has been.  Once
+implemented, delete this comment.
+
+### The Fees Program
+
+The Solana cluster has an on-chain program called Fees with just one global
+account called the *fees account*. The account holds a set of fee parameters
+that the cluster will use to calculate transaction fees in the following epoch.
+Any slot leader may submit congestion statistics to the fees account and the
+Fees program uses that data to automatically adjust the fees to maximize
+resource usage.  Each validator checks congestion statistics and rejects blocks
+that include congestion data inconsistent with its own observations.
+
+The client uses the JSON RPC API to query the cluster for the current fee
+parameters.  Those parameters are tagged with a blockhash and remain valid
+until that blockhash is old enough to be rejected by the slot leader.
+
+### Calculating fees
+
+The fees account contains a variety of fee parameters. Before sending a
+transaction to the cluster, a client may submit the transaction and fee account
+data to an SDK module called the *fee calculator*. So long as the client's SDK
```

Comment:
This automatic fee calibration is super hard to pull off without an attack vector. 

I would initially just have the fee set by the validators in their VoteState account and compute the median at every epoch.

Reply:
Yeah, good point. It's subject to an eclipse attack to DoS the cluster. How about validators only upload congestion data into their vote accounts? And then the fees are derived from stake-weighted averages?

Reply:
@ericlwilliams, can you chime in.  This would be our first foray into on-chain governance.  What would be the least worst way to do this?

Reply:
why do the validators need to submit network congestion data, can't this be calculated from the ledger? Or is this assuming that the network is so congested not everything is making it into the ledger - in which case I wonder if this is worth trying to solve now (i.e. governance) and use early data to see if/when congestion will actually be an issue (and how stakeholder relationships play out).
Another concern I have is how the replicators play in this. A fixed percentage of each transaction fee is going to have to go to the mining pool (essentially, to the replicators), which means they have interest in how this fee is set too. Also, the relationship between PoRep transaction fees and non-PoRep transaction fees is something the network should pay attention to, as it determines the long-term sustainability of the replication economy (at least, as the design now stands). It's unclear at this point how replicators will participate and I think we may want to wait to see how they integrate with the validator system to understand the dynamics well enough to be able to account for them as well

Reply:
If I'm understanding the proposal, this doesn't seem exactly like on-chain governance, rather more like an algorithmic determination, unless the wiggle-room that the validators have in what they do (or don't) submit is considered a form of voting.  
If we want to ignore replicators for now, this approach seems okay except for my question above regarding the need for validators to actually submit this data.  Could an mvp be that this is just calculated from the observed transaction history?

Reply:
I think the on-chain governance part setting the cluster-wide TPS capacity.  Manipulating that would be a centralization technique that governance could guard against.

Reply:
But isn't that determined by the congestion statistics that are submitted (per your comment below about low vs high bandwidth validators)?  Or are you proposing an actual vote on cluster-wide TPS?   For the former, I'm not following exactly how that is on-chain governance, vs an on-chain algorithmic adjustment

Reply:
I'm proposing that *TPS capacity* must be voted on. For the short-term, we can add it to the genesis block.

Reply:
Maybe we're using different terminology.  What I'm reading in this proposal is validators submitting congestion data, which leads to a fee adjustment/determination, which is likely to effect the network traffic (TPS).
In this proposal I don't see anything about TPS capacity being voted on directly - is that outside of this scope, or are you referring to the fee adjustment mechanism as voting on TPS capacity?

Reply:
I hit the reset button. Check out the latest proposal.  The things that will eventually be voted on is any parameter in the genesis block. In this proposal, that's just *SPS target*, a term defined by me a few minutes ago.

Reply:
Ah cool. So just to be clear, the governance process to vote on SPS target is outside the scope of this proposal. This just covers how validators submission of ledger data SPS adjusts the fee based on the current SPS target.

Reply:
Yes, I think Anatoly brought you in here to offer a concrete example of where we'd use on-chain governance. If we already had a good idea of how we wanted to do that,  we could tack it onto this proposal, but dealing with it separately sounds great to me.